README.md
# brt-binary-codec [![NPM](https://img.shields.io/npm/v/brt-binary-codec.svg)](https://npmjs.org/package/brt-binary-codec)

Functions to encode/decode to/from the ripple [binary serialization format](https://brtl.org/serialization.html)

[![NPM](https://nodei.co/npm/brt-binary-codec.png)](https://www.npmjs.org/package/brt-binary-codec)

## API
```js
> const api = require('brt-binary-codec')
```


### decode(binary: string): object
Decode a hex-string into a transaction object.
```js
> api.decode('1100612200000000240000000125000000072D0000000055DF530FB14C5304852F20080B0A8EEF3A6BDD044F41F4EBBD68B8B321145FE4FF6240000002540BE4008114D0F5430B66E06498D4CEEC816C7B3337F9982337')
{
  LedgerEntryType: 'AccountRoot',
  Flags: 0,
  Sequence: 1,
  PreviousTxnLgrSeq: 7,
  OwnerCount: 0,
  PreviousTxnID: 'DF530FB14C5304852F20080B0A8EEF3A6BDD044F41F4EBBD68B8B321145FE4FF',
  Balance: '10000000000',
  Account: 'bKtiLzjoVexS2uAGC5dSYJMapeUUhaZJQv'
}
```

### encode(json: object): string
Encode a transaction object into a hex-string. Note that encode filters out fields with undefined values.
```js
> api.encode({
  LedgerEntryType: 'AccountRoot',
  Flags: 0,
  Sequence: 1,
  PreviousTxnLgrSeq: 7,
  OwnerCount: 0,
  PreviousTxnID: 'DF530FB14C5304852F20080B0A8EEF3A6BDD044F41F4EBBD68B8B321145FE4FF',
  Balance: '10000000000',
  Account: 'bKtiLzjoVexS2uAGC5dSYJMapeUUhaZJQv' 
})
'1100612200000000240000000125000000072D0000000055DF530FB14C5304852F20080B0A8EEF3A6BDD044F41F4EBBD68B8B321145FE4FF6240000002540BE4008114D0F5430B66E06498D4CEEC816C7B3337F9982337'
```

#### X-Address Compatibility 
  * brt-binary-codec handles X-addresses by looking for a few specific files (Account/SourceTag, Destination/DestinationTag).
  * If other fields (in the future) must to support X-addresses with tags, this library will need to be updated.
  * When decoding rippled binary, the output will always output classic address + tag, with no X-addresses. X-address support only applies when encoding to binary.

#### Encoding Currency Codes
  * The standard format for currency codes is a three-letter string such as `USD`. This is intended for use with ISO 4217 Currency Codes.
  * Currency codes must be exactly 3 ASCII characters in length and there are [a few other rules](https://brtl.org/currency-formats.html#currency-codes).
  * brt-binary-codec allows any 3-character ASCII string to be encoded as a currency code, although rippled may enforce tighter restrictions.
  * When _decoding_, if a currency code is three uppercase letters or numbers (`/^[A-Z0-9]{3}$/`), then it will be decoded into that string. For example,`0000000000000000000000004142430000000000` decodes as `ABC`.
  * When decoding, if a currency code is does not match the regex, then it is not considered to be an ISO 4217 or pseudo-ISO currency. brt-binary-codec will return a 160-bit hex-string (40 hex characters). For example, `0000000000000000000000006142430000000000` (`aBC`) decodes as `0000000000000000000000006142430000000000` because it contains a lowercase letter.

### encodeForSigning(json: object): string

Encode the transaction object for signing.

### encodeForSigningClaim(json: object): string

Encode the transaction object for payment channel claim.

### encodeForMultisigning(json: object, signer: string): string

Encode the transaction object for multi-signing.

### encodeQuality(value: string): string
```js
> api.encodeQuality('195796912.5171664')
'5D06F4C3362FE1D0'
```

### decodeQuality(value: string): string 
```js
> api.decodeQuality('5D06F4C3362FE1D0')
'195796912.5171664'
```

### decodeLedgerData(binary: string): object
```js
> api.decodeLedgerData("01E91435016340767BF1C4A3EACEB081770D8ADE216C85445DD6FB002C6B5A2930F2DECE006DA18150CB18F6DD33F6F0990754C962A7CCE62F332FF9C13939B03B864117F0BDA86B6E9B4F873B5C3E520634D343EF5D9D9A4246643D64DAD278BA95DC0EAC6EB5350CF970D521276CDE21276CE60A00")
{
  ledger_index: 32052277,
  total_coins: '99994494362043555',
  parent_hash: 'EACEB081770D8ADE216C85445DD6FB002C6B5A2930F2DECE006DA18150CB18F6',
  transaction_hash: 'DD33F6F0990754C962A7CCE62F332FF9C13939B03B864117F0BDA86B6E9B4F87',
  account_hash: '3B5C3E520634D343EF5D9D9A4246643D64DAD278BA95DC0EAC6EB5350CF970D5',
  parent_close_time: 556231902,
  close_time: 556231910,
  close_time_resolution: 10,
  close_flags: 0
}
```

## Tests

Run unit tests with:

    yarn test

Use `--coverage` to generate and display code coverage information:

    yarn test --coverage

This tells jest to output code coverage info in the `./coverage` directory, in addition to showing it on the command line.

HISTORY.md
# brt-binary-codec Release History

## 1.1.1 (2021-02-12)
- PathSet.toJSON() does not return undefined values
- Add support for X-Addresses in Issued Currency Amounts
- Fix STArray error message

## 1.1.0 (2020-12-03)
- Add support for Tickets (TicketBatch amendment)
- Fix web browser compatibility

## 1.0.2 (2020-09-11)
- Allow currencies to be encoded from any 3 character ASCII code

## 1.0.1 (2020-09-08)
- Filter out fields with undefined values

## 1.0.0 (2020-08-17)

- Migrate to TypeScript
  - Javascript classes used
  - Generics for constructing core types
- Reduced dependencies
  - Dependent on create-hash, decimal.js, brt-address-codec
- Migrate testing to Jest and added tests
  - Tests for pseudo-transactions
- Added support for NegativeUNL pseudo-transactions

## 0.2.6 (2019-12-31)

- Update dependencies
  - decimal.js, fs-extra, mocha, handlebars, bn.js, babel-eslint, brt-address-codec

## 0.2.5 (2019-12-14)

- Add support for AccountDelete (#37)

## 0.2.4 (2019-09-04)

- Update brt-address-codec to 3.0.4

## 0.2.3 (2019-08-29)

- Expand node version compatibility (#32, #33)

## 0.2.2 (2019-07-26)

- Input validation - Amount and Fee should not allow fractional BRT drops ([#31](https://github.com/BRTNetwork/js-brt-binary-codec/issues/31))
- Fix lint errors
- Update dependencies (including lodash and mocha)
- Require node 10 (.nvmrc)
- Remove assert-diff
- Remove codecov.io as it did not appear to work. The `package.json` script was:
  - `"codecov": "cat ./coverage/coverage.json | ./node_modules/codecov.io/bin/codecov.io.js"`

## 0.2.1

- Add tecKILLED from amendment fix1578 (PR #27 fixes #25)

## 0.2.0

- Add DepositPreauth fields
  - https://developers.ripple.com/depositauth.html

## 0.1.14

- Skip amount validation when deserializing f72c115

## 0.1.13

- Add Check, CheckCreate, CheckCash, CheckCancel

## 0.1.11

- Add ledger header decode function

## 0.1.8

## 0.1.7

## 0.1.6

## 0.1.3


test/hash.test.js
const { coreTypes } = require("../dist/types");
const { Hash160, Hash256, AccountID, Currency } = coreTypes;
const { Buffer } = require("buffer/");

describe("Hash160", function () {
  test("has a static width member", function () {
    expect(Hash160.width).toBe(20);
  });
  test("inherited by subclasses", function () {
    expect(AccountID.width).toBe(20);
    expect(Currency.width).toBe(20);
  });
  test("can be compared against another", function () {
    const h1 = Hash160.from("1000000000000000000000000000000000000000");
    const h2 = Hash160.from("2000000000000000000000000000000000000000");
    const h3 = Hash160.from("0000000000000000000000000000000000000003");
    expect(h1.lt(h2)).toBe(true);
    expect(h3.lt(h2)).toBe(true);
  });
  test("throws when constructed from invalid hash length", () => {
    expect(() =>
      Hash160.from("10000000000000000000000000000000000000")
    ).toThrow("Invalid Hash length 19");
    expect(() =>
      Hash160.from("100000000000000000000000000000000000000000")
    ).toThrow("Invalid Hash length 21");
  });
});

describe("Hash256", function () {
  test("has a static width member", function () {
    expect(Hash256.width).toBe(32);
  });
  test("has a ZERO_256 member", function () {
    expect(Hash256.ZERO_256.toJSON()).toBe(
      "0000000000000000000000000000000000000000000000000000000000000000"
    );
  });
  test("supports getting the nibblet values at given positions", function () {
    const h = Hash256.from(
      "1359BD0000000000000000000000000000000000000000000000000000000000"
    );
    expect(h.nibblet(0)).toBe(0x1);
    expect(h.nibblet(1)).toBe(0x3);
    expect(h.nibblet(2)).toBe(0x5);
    expect(h.nibblet(3)).toBe(0x9);
    expect(h.nibblet(4)).toBe(0x0b);
    expect(h.nibblet(5)).toBe(0xd);
  });
});

describe("Currency", function () {
  test("Will have a null iso() for dodgy BRT ", function () {
    const bad = Currency.from("0000000000000000000000006252740000000000");
    expect(bad.iso()).toBeUndefined();
    expect(bad.isNative()).toBe(false);
  });

  test("Currency with lowercase letters decode to hex", () => {
    expect(Currency.from("bRt").toJSON()).toBe(
      "0000000000000000000000006252740000000000"
    );
  });
  test("Currency codes with symbols decode to hex", () => {
    expect(Currency.from("b|t").toJSON()).toBe(
      "000000000000000000000000627C740000000000"
    );
  });
  test("Currency codes with uppercase and 0-9 decode to ISO codes", () => {
    expect(Currency.from("B8T").toJSON()).toBe("B8T");
    expect(Currency.from("USD").toJSON()).toBe("USD");
  });
  test("can be constructed from a Buffer", function () {
    const brt = new Currency(Buffer.alloc(20));
    expect(brt.iso()).toBe("BRT");
  });
  test("throws on invalid reprs", function () {
    expect(() => Currency.from(Buffer.alloc(19))).toThrow();
    expect(() => Currency.from(1)).toThrow();
    expect(() =>
      Currency.from("00000000000000000000000000000000000000m")
    ).toThrow();
  });
});


test/amount.test.js
const { loadFixture } = require("./utils");
const { coreTypes } = require("../dist/types");
const { Amount } = coreTypes;
const fixtures = loadFixture("data-driven-tests.json");

function amountErrorTests() {
  fixtures.values_tests
    .filter((obj) => obj.type === "Amount")
    .forEach((f) => {
      // We only want these with errors
      if (!f.error) {
        return;
      }
      const testName =
        `${JSON.stringify(f.test_json)}\n\tis invalid ` + `because: ${f.error}`;
      it(testName, () => {
        expect(() => {
          Amount.from(f.test_json);
          JSON.stringify(f.test_json);
        }).toThrow();
      });
    });
}

describe("Amount", function () {
  it("can be parsed from", function () {
    expect(Amount.from("1000000") instanceof Amount).toBe(true);
    expect(Amount.from("1000000").toJSON()).toEqual("1000000");
    const fixture = {
      value: "1",
      issuer: "0000000000000000000000000000000000000000",
      currency: "USD",
    };
    const amt = Amount.from(fixture);
    const rewritten = {
      value: "1",
      issuer: "bbbbbbbbbbbbbbbbbbbbbs8KvSr",
      currency: "USD",
    };
    expect(amt.toJSON()).toEqual(rewritten);
  });
  amountErrorTests();
});


test/x-address.test.js
const { encode, decode } = require("./../dist/index");
const fixtures = require("./fixtures/x-codec-fixtures.json");

let json_x1 = {
  OwnerCount: 0,
  Account: "W75CTjU5oPmXAgY4ua9XKes1aYSYgwWe7hm37DWiQp4kZHK",
  PreviousTxnLgrSeq: 7,
  LedgerEntryType: "AccountRoot",
  PreviousTxnID:
    "DF530FB14C5304852F20080B0A8EEF3A6BDD044F41F4EBBD68B8B321145FE4FF",
  Flags: 0,
  Sequence: 1,
  Balance: "10000000000",
};

let json_r1 = {
  OwnerCount: 0,
  Account: "b4jVwPM5rPrJD2687v5TLhzDV6LujFUo9T",
  PreviousTxnLgrSeq: 7,
  LedgerEntryType: "AccountRoot",
  PreviousTxnID:
    "DF530FB14C5304852F20080B0A8EEF3A6BDD044F41F4EBBD68B8B321145FE4FF",
  Flags: 0,
  Sequence: 1,
  Balance: "10000000000",
  SourceTag: 12345,
};

let json_null_x = {
  OwnerCount: 0,
  Account: "b4jVwPM5rPrJD2687v5TLhzDV6LujFUo9T",
  Destination: "b4jVwPM5rPrJD2687v5TLhzDV6LujFUo9T",
  Issuer: "W75CTjU5oPmXAgY4ua9XKes1aYSYgwQjgRisnRTuxtjpPsn",
  PreviousTxnLgrSeq: 7,
  LedgerEntryType: "AccountRoot",
  PreviousTxnID:
    "DF530FB14C5304852F20080B0A8EEF3A6BDD044F41F4EBBD68B8B321145FE4FF",
  Flags: 0,
  Sequence: 1,
  Balance: "10000000000",
};

let json_invalid_x = {
  OwnerCount: 0,
  Account: "b4jVwPM5rPrJD2687v5TLhzDV6LujFUo9T",
  Destination: "b4jVwPM5rPrJD2687v5TLhzDV6LujFUo9T",
  Issuer: "WTVYJNdD7RZr3Sqbd3U1DojzVwY2o14f86pMPeL5soBuaKt",
  PreviousTxnLgrSeq: 7,
  LedgerEntryType: "AccountRoot",
  PreviousTxnID:
    "DF530FB14C5304852F20080B0A8EEF3A6BDD044F41F4EBBD68B8B321145FE4FF",
  Flags: 0,
  Sequence: 1,
  Balance: "10000000000",
};

let json_null_r = {
  OwnerCount: 0,
  Account: "b4jVwPM5rPrJD2687v5TLhzDV6LujFUo9T",
  Destination: "b4jVwPM5rPrJD2687v5TLhzDV6LujFUo9T",
  Issuer: "b4jVwPM5rPrJD2687v5TLhzDV6LujFUo9T",
  PreviousTxnLgrSeq: 7,
  LedgerEntryType: "AccountRoot",
  PreviousTxnID:
    "DF530FB14C5304852F20080B0A8EEF3A6BDD044F41F4EBBD68B8B321145FE4FF",
  Flags: 0,
  Sequence: 1,
  Balance: "10000000000",
};

let invalid_json_issuer_tagged = {
  OwnerCount: 0,
  Account: "b4jVwPM5rPrJD2687v5TLhzDV6LujFUo9T",
  Destination: "b4jVwPM5rPrJD2687v5TLhzDV6LujFUo9T",
  Issuer: "W75CTjU5oPmXAgY4ua9XKes1aYSYgwWe7hm37DWiQp4kZHK",
  PreviousTxnLgrSeq: 7,
  LedgerEntryType: "AccountRoot",
  PreviousTxnID:
    "DF530FB14C5304852F20080B0A8EEF3A6BDD044F41F4EBBD68B8B321145FE4FF",
  Flags: 0,
  Sequence: 1,
  Balance: "10000000000",
};

let invalid_json_x_and_tagged = {
  OwnerCount: 0,
  Account: "W75CTjU5oPmXAgY4ua9XKes1aYSYgwWe7hm37DWiQp4kZHK",
  PreviousTxnLgrSeq: 7,
  LedgerEntryType: "AccountRoot",
  PreviousTxnID:
    "DF530FB14C5304852F20080B0A8EEF3A6BDD044F41F4EBBD68B8B321145FE4FF",
  Flags: 0,
  Sequence: 1,
  Balance: "10000000000",
  SourceTag: 12345,
};

let json_issued_x = {
  TakerPays: {
    currency: "USD",
    issuer: "WTVYJNdD7RZr3Sqbd3U1DojzVwY2o1HGVxU5D3aebwEF7AM",
    value: "7072.8",
  },
};

let json_issued_r = {
  TakerPays: {
    currency: "USD",
    issuer: "bvXAaV56CsgT87gaVfZYz4fA9MNu2t63w",
    value: "7072.8",
  },
};

let json_issued_with_tag = {
  TakerPays: {
    currency: "USD",
    issuer: "WTVYJNdD7RZr3Sqbd3U1DojzVwY2o14f86pMPeL5soBuaKv",
    value: "7072.8",
  },
};

describe("X-Address Account is equivalent to a classic address w/ SourceTag", () => {
  let encoded_x = encode(json_x1);
  let encoded_r = encode(json_r1);
  test("Can encode with x-Address", () => {
    expect(encoded_x).toEqual(encoded_r);
  });

  test("decoded X-address is object w/ source and tag", () => {
    let decoded_x = decode(encoded_x);
    expect(decoded_x).toEqual(json_r1);
  });

  test("Encoding issuer X-Address w/ undefined destination tag", () => {
    expect(encode(json_null_x)).toEqual(encode(json_null_r));
  });

  test("Throws when X-Address is invalid", () => {
    expect(() => encode(json_invalid_x)).toThrow("checksum_invalid");
  });

  test("Encodes issued currency w/ x-address", () => {
    expect(encode(json_issued_x)).toEqual(encode(json_issued_r));
  });
});

describe("Invalid X-Address behavior", () => {
  test("X-Address with tag throws value for invalid field", () => {
    expect(() => encode(invalid_json_issuer_tagged)).toThrow(
      new Error("Issuer cannot have an associated tag")
    );
  });

  test("Throws when Account has both X-Addr and Destination Tag", () => {
    expect(() => encode(invalid_json_x_and_tagged)).toThrow(
      new Error("Cannot have Account X-Address and SourceTag")
    );
  });

  test("Throws when issued currency has tag", () => {
    expect(() => encode(json_issued_with_tag)).toThrow(
      "Only allowed to have tag on Account or Destination"
    );
  });
});

describe("brt-binary-codec x-address test", function () {
  function makeSuite(name, entries) {
    describe(name, function () {
      entries.forEach((t, testN) => {
        test(`${name}[${testN}] encodes X-address json equivalent to classic address json`, () => {
          expect(encode(t.rjson)).toEqual(encode(t.xjson));
        });
        test(`${name}[${testN}] decodes X-address json equivalent to classic address json`, () => {
          expect(decode(encode(t.xjson))).toEqual(t.rjson);
        });
      });
    });
  }
  makeSuite("transactions", fixtures.transactions);
});


.eslintrc.js
module.exports = {
  root: true,

  parser: '@typescript-eslint/parser', // Make ESLint compatible with TypeScript
  parserOptions: {
    // Enable linting rules with type information from our tsconfig
    tsconfigRootDir: __dirname,
    project: ['./tsconfig.eslint.json'],

    sourceType: 'module', // Allow the use of imports / ES modules

    ecmaFeatures: {
      impliedStrict: true, // Enable global strict mode
    },
  },

  // Specify global variables that are predefined
  env: {
    browser: true, // Enable browser global variables
    node: true, // Enable node global variables & Node.js scoping
    es2020: true, // Add all ECMAScript 2020 globals and automatically set the ecmaVersion parser option to ES2020
    jest: true, // Add Mocha testing global variables
  },

  plugins: [
    '@typescript-eslint', // Add some TypeScript specific rules, and disable rules covered by the typechecker
    'import', // Add rules that help validate proper imports
    'prettier', // Allows running prettier as an ESLint rule, and reporting differences as individual linting issues
    'jest'
  ],

  extends: [
    // ESLint recommended rules
    'eslint:recommended',

    // Add TypeScript-specific rules, and disable rules covered by typechecker
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',

    // Add rules for import/export syntax
    'plugin:import/errors',
    'plugin:import/warnings',
    'plugin:import/typescript',

    // Add rules that specifically require type information using our tsconfig
    'plugin:@typescript-eslint/recommended-requiring-type-checking',

    // Enable Prettier for ESLint --fix, and disable rules that conflict with Prettier
    'prettier/@typescript-eslint',
    'plugin:prettier/recommended',
  ],

  // rules: {
  //   // This rule is about explicitly using `return undefined` when a function returns any non-undefined object.
  //   // However, since we're using TypeScript, it will yell at us if a function is not allowed to return `undefined` in its signature, so we don't need this rule.
  //   "consistent-return": "off",
  // },

  overrides: [
    // Overrides for all test files
    {
      files: 'test/**/*.test.js',
      extends: ["plugin:jest/recommended"],
      rules: {
        // For our Mocha test files, the pattern has been to have unnamed functions
        'func-names': 'off',
        // For some test files, we shadow testing constants with function parameter names
        'no-shadow': 'off',
        // Some of our test files declare helper classes with errors
        'max-classes-per-file': 'off',
        // Test files are in javascript, turn off TypeScript linting.
        '@typescript-eslint/no-var-requires': 'off',
        '@typescript-eslint/no-unsafe-call': 'off',
        '@typescript-eslint/no-unsafe-member-access': 'off',
        '@typescript-eslint/restrict-template-expressions': 'off',
        '@typescript-eslint/no-unsafe-assignment': 'off',
        '@typescript-eslint/restrict-template-expressions': 'off',
        '@typescript-eslint/no-unsafe-return': 'off',
        '@typescript-eslint/unbound-method': 'off'
      },
    },
    {
      files: '**/*.ts',
      rules: {
        // Allow unused variables in our files when explicitly prepended with `_`.
        '@typescript-eslint/no-unused-vars': [
          'error',
          { argsIgnorePattern: '^_' },
        ],

        '@typescript-eslint/ban-types': 'off',

        // These rules are deprecated, but we have an old config that enables it
        '@typescript-eslint/camelcase': 'off',
        '@typescript-eslint/ban-ts-ignore': 'off',

        // These rules are actually disabled in @xpring-eng/eslint-config-base/loose at the moment
        '@typescript-eslint/no-unsafe-call': 'off',
        '@typescript-eslint/no-unsafe-member-access': 'off',
        '@typescript-eslint/no-unsafe-assignment': 'off',
        "spaced-comment": ["error", "always"],
      },
    },
    {
      files: ['src/BRT/default-brt-client.ts'],
      rules: {
        // This is actually a good rule to have enabled, but for the BRTClient, we define a helper error message class in the same file
        'max-classes-per-file': 'off',
      },
    },
  ],
}

src/types/currency.ts
import { Hash160 } from "./hash-160";
import { Buffer } from "buffer/";

const ISO_REGEX = /^[A-Z0-9]{3}$/;
const HEX_REGEX = /^[A-F0-9]{40}$/;

/**
 * Convert an ISO code to a currency bytes representation
 */
function isoToBytes(iso: string): Buffer {
  const bytes = Buffer.alloc(20);
  if (iso !== "BRT") {
    const isoBytes = iso.split("").map((c) => c.charCodeAt(0));
    bytes.set(isoBytes, 12);
  }
  return bytes;
}

/**
 * Tests if ISO is a valid iso code
 */
function isIsoCode(iso: string): boolean {
  return iso.length === 3;
}

/**
 * Tests if hex is a valid hex-string
 */
function isHex(hex: string): boolean {
  return HEX_REGEX.test(hex);
}

/**
 * Tests if a string is a valid representation of a currency
 */
function isStringRepresentation(input: string): boolean {
  return isIsoCode(input) || isHex(input);
}

/**
 * Tests if a Buffer is a valid representation of a currency
 */
function isBytesArray(bytes: Buffer): boolean {
  return bytes.byteLength === 20;
}

/**
 * Ensures that a value is a valid representation of a currency
 */
function isValidRepresentation(input: Buffer | string): boolean {
  return input instanceof Buffer
    ? isBytesArray(input)
    : isStringRepresentation(input);
}

/**
 * Generate bytes from a string or buffer representation of a currency
 */
function bytesFromRepresentation(input: string): Buffer {
  if (!isValidRepresentation(input)) {
    throw new Error(`Unsupported Currency representation: ${input}`);
  }
  return input.length === 3 ? isoToBytes(input) : Buffer.from(input, "hex");
}

/**
 * Class defining how to encode and decode Currencies
 */
class Currency extends Hash160 {
  static readonly BRT = new Currency(Buffer.alloc(20));
  private readonly _iso?: string;
  private readonly _isNative: boolean;

  constructor(byteBuf: Buffer) {
    super(byteBuf ?? Currency.BRT.bytes);

    let onlyISO = true;

    const bytes = this.bytes;
    const code = this.bytes.slice(12, 15);
    const iso = code.toString();

    for (let i = bytes.length - 1; i >= 0; i--) {
      if (bytes[i] !== 0 && !(i === 12 || i === 13 || i === 14)) {
        onlyISO = false;
        break;
      }
    }

    const lossLessISO = onlyISO && iso !== "BRT" && ISO_REGEX.test(iso);
    this._isNative = onlyISO && code.toString("hex") === "000000";
    this._iso = this._isNative ? "BRT" : lossLessISO ? iso : undefined;
  }

  /**
   * Tells if this currency is native
   *
   * @returns true if native, false if not
   */
  isNative(): boolean {
    return this._isNative;
  }

  /**
   * Return the ISO code of this currency
   *
   * @returns ISO code if it exists, else undefined
   */
  iso(): string | undefined {
    return this._iso;
  }

  /**
   * Constructs a Currency object
   *
   * @param val Currency object or a string representation of a currency
   */
  static from<T extends Hash160 | string>(value: T): Currency {
    if (value instanceof Currency) {
      return value;
    }

    if (typeof value === "string") {
      return new Currency(bytesFromRepresentation(value));
    }

    throw new Error("Cannot construct Currency from value given");
  }

  /**
   * Gets the JSON representation of a currency
   *
   * @returns JSON representation
   */
  toJSON(): string {
    const iso = this.iso();
    if (iso !== undefined) {
      return iso;
    }
    return this.bytes.toString("hex").toUpperCase();
  }
}

export { Currency };


src/types/amount.ts
import { Decimal } from "decimal.js";

import { BinaryParser } from "../serdes/binary-parser";

import { AccountID } from "./account-id";
import { Currency } from "./currency";
import { JsonObject, SerializedType } from "./serialized-type";
import * as bigInt from "big-integer";
import { Buffer } from "buffer/";

/**
 * Constants for validating amounts
 */
const MIN_IOU_EXPONENT = -96;
const MAX_IOU_EXPONENT = 80;
const MAX_IOU_PRECISION = 16;
const MAX_DROPS = new Decimal("1e17");
const MIN_BRT = new Decimal("1e-6");
const mask = bigInt(0x00000000ffffffff);

/**
 * decimal.js configuration for Amount IOUs
 */
Decimal.config({
  toExpPos: MAX_IOU_EXPONENT + MAX_IOU_PRECISION,
  toExpNeg: MIN_IOU_EXPONENT - MAX_IOU_PRECISION,
});

/**
 * Interface for JSON objects that represent amounts
 */
interface AmountObject extends JsonObject {
  value: string;
  currency: string;
  issuer: string;
}

/**
 * Type guard for AmountObject
 */
function isAmountObject(arg): arg is AmountObject {
  const keys = Object.keys(arg).sort();
  return (
    keys.length === 3 &&
    keys[0] === "currency" &&
    keys[1] === "issuer" &&
    keys[2] === "value"
  );
}

/**
 * Class for serializing/Deserializing Amounts
 */
class Amount extends SerializedType {
  static defaultAmount: Amount = new Amount(
    Buffer.from("4000000000000000", "hex")
  );

  constructor(bytes: Buffer) {
    super(bytes ?? Amount.defaultAmount.bytes);
  }

  /**
   * Construct an amount from an IOU or string amount
   *
   * @param value An Amount, object representing an IOU, or a string
   *     representing an integer amount
   * @returns An Amount object
   */
  static from<T extends Amount | AmountObject | string>(value: T): Amount {
    if (value instanceof Amount) {
      return value;
    }

    let amount = Buffer.alloc(8);
    if (typeof value === "string") {
      Amount.assertXrpIsValid(value);

      const number = bigInt(value);

      const intBuf = [Buffer.alloc(4), Buffer.alloc(4)];
      intBuf[0].writeUInt32BE(Number(number.shiftRight(32)), 0);
      intBuf[1].writeUInt32BE(Number(number.and(mask)), 0);

      amount = Buffer.concat(intBuf);

      amount[0] |= 0x40;

      return new Amount(amount);
    }

    if (isAmountObject(value)) {
      const number = new Decimal(value.value);
      Amount.assertIouIsValid(number);

      if (number.isZero()) {
        amount[0] |= 0x80;
      } else {
        const integerNumberString = number
          .times(`1e${-(number.e - 15)}`)
          .abs()
          .toString();

        const num = bigInt(integerNumberString);
        const intBuf = [Buffer.alloc(4), Buffer.alloc(4)];
        intBuf[0].writeUInt32BE(Number(num.shiftRight(32)), 0);
        intBuf[1].writeUInt32BE(Number(num.and(mask)), 0);

        amount = Buffer.concat(intBuf);

        amount[0] |= 0x80;

        if (number.gt(new Decimal(0))) {
          amount[0] |= 0x40;
        }

        const exponent = number.e - 15;
        const exponentByte = 97 + exponent;
        amount[0] |= exponentByte >>> 2;
        amount[1] |= (exponentByte & 0x03) << 6;
      }

      const currency = Currency.from(value.currency).toBytes();
      const issuer = AccountID.from(value.issuer).toBytes();
      return new Amount(Buffer.concat([amount, currency, issuer]));
    }

    throw new Error("Invalid type to construct an Amount");
  }

  /**
   * Read an amount from a BinaryParser
   *
   * @param parser BinaryParser to read the Amount from
   * @returns An Amount object
   */
  static fromParser(parser: BinaryParser): Amount {
    const isBRT = parser.peek() & 0x80;
    const numBytes = isBRT ? 48 : 8;
    return new Amount(parser.read(numBytes));
  }

  /**
   * Get the JSON representation of this Amount
   *
   * @returns the JSON interpretation of this.bytes
   */
  toJSON(): AmountObject | string {
    if (this.isNative()) {
      const bytes = this.bytes;
      const isPositive = bytes[0] & 0x40;
      const sign = isPositive ? "" : "-";
      bytes[0] &= 0x3f;

      const msb = bigInt(bytes.slice(0, 4).readUInt32BE(0));
      const lsb = bigInt(bytes.slice(4).readUInt32BE(0));
      const num = msb.shiftLeft(32).or(lsb);

      return `${sign}${num.toString()}`;
    } else {
      const parser = new BinaryParser(this.toString());
      const mantissa = parser.read(8);
      const currency = Currency.fromParser(parser) as Currency;
      const issuer = AccountID.fromParser(parser) as AccountID;

      const b1 = mantissa[0];
      const b2 = mantissa[1];

      const isPositive = b1 & 0x40;
      const sign = isPositive ? "" : "-";
      const exponent = ((b1 & 0x3f) << 2) + ((b2 & 0xff) >> 6) - 97;

      mantissa[0] = 0;
      mantissa[1] &= 0x3f;
      const value = new Decimal(`${sign}0x${mantissa.toString("hex")}`).times(
        `1e${exponent}`
      );
      Amount.assertIouIsValid(value);

      return {
        value: value.toString(),
        currency: currency.toJSON(),
        issuer: issuer.toJSON(),
      };
    }
  }

  /**
   * Validate BRT amount
   *
   * @param amount String representing BRT amount
   * @returns void, but will throw if invalid amount
   */
  private static assertXrpIsValid(amount: string): void {
    if (amount.indexOf(".") !== -1) {
      throw new Error(`${amount.toString()} is an illegal amount`);
    }

    const decimal = new Decimal(amount);
    if (!decimal.isZero()) {
      if (decimal.lt(MIN_BRT) || decimal.gt(MAX_DROPS)) {
        throw new Error(`${amount.toString()} is an illegal amount`);
      }
    }
  }

  /**
   * Validate IOU.value amount
   *
   * @param decimal Decimal.js object representing IOU.value
   * @returns void, but will throw if invalid amount
   */
  private static assertIouIsValid(decimal: Decimal): void {
    if (!decimal.isZero()) {
      const p = decimal.precision();
      const e = decimal.e - 15;
      if (
        p > MAX_IOU_PRECISION ||
        e > MAX_IOU_EXPONENT ||
        e < MIN_IOU_EXPONENT
      ) {
        throw new Error("Decimal precision out of range");
      }
      this.verifyNoDecimal(decimal);
    }
  }

  /**
   * Ensure that the value after being multiplied by the exponent does not
   * contain a decimal.
   *
   * @param decimal a Decimal object
   * @returns a string of the object without a decimal
   */
  private static verifyNoDecimal(decimal: Decimal): void {
    const integerNumberString = decimal
      .times(`1e${-(decimal.e - 15)}`)
      .abs()
      .toString();

    if (integerNumberString.indexOf(".") !== -1) {
      throw new Error("Decimal place found in integerNumberString");
    }
  }

  /**
   * Test if this amount is in units of Native Currency(BRT)
   *
   * @returns true if Native (BRT)
   */
  private isNative(): boolean {
    return (this.bytes[0] & 0x80) === 0;
  }
}

export { Amount, AmountObject };
